cmd/sourcemaps/upload.go

```diff
@@ -0,0 +1,334 @@
+		if err != nil {
+			return err
+		}
+
+		if d.IsDir() && d.Name() == "node_modules" {
+			return filepath.SkipDir
+		}
+
+		if !d.IsDir() && (strings.HasSuffix(filePath, ".js.map") || strings.HasSuffix(filePath, ".js")) {
+			relPath, err := filepath.Rel(path, filePath)
+			if err != nil {
+				return err
+			}
+
+			files = append(files, SourceMapFile{
+				Path: filePath,
+				Name: relPath,
+			})
+
+			routeGroupRemovedPath := routeGroupPattern.ReplaceAllString(relPath, "")
+			if routeGroupRemovedPath != relPath {
+				files = append(files, SourceMapFile{
+					Path: filePath,
+					Name: routeGroupRemovedPath,
+				})
+			}
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(files) == 0 {
+		return nil, fmt.Errorf("no .js.map files found. Please double check that you have generated sourcemaps for your app")
+	}
+
+	return files, nil
+}
+
+func getS3Key(organizationID, version, basePath, fileName string) string {
+	if version == "" {
+		version = "unversioned"
+	}
+
+	if basePath != "" && !strings.HasSuffix(basePath, "/") {
+		basePath = basePath + "/"
+	}
+
+	return fmt.Sprintf("%s/%s/%s%s", organizationID, version, basePath, fileName)
+}
+
+func getSourceMapUploadUrls(apiKey string, paths []string, backendUrl string) ([]string, error) {
+	variables := map[string]interface{}{
+		"api_key": apiKey,
+		"paths":   paths,
+	}
+
+	reqBody, err := json.Marshal(map[string]interface{}{
+		"query":     getSourceMapUrlsQuery,
+		"variables": variables,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest("POST", backendUrl, bytes.NewBuffer(reqBody))
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var urlsResp SourceMapUrlsResponse
+	if err := json.Unmarshal(body, &urlsResp); err != nil {
+		return nil, err
+	}
+
+	if len(urlsResp.Data.GetSourceMapUploadUrls) == 0 {
+		return nil, fmt.Errorf("unable to generate source map upload urls")
+	}
+
+	return urlsResp.Data.GetSourceMapUploadUrls, nil
+}
+
+func uploadFile(filePath, uploadUrl, name string) error {
+	fileContent, err := os.ReadFile(filePath)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest("PUT", uploadUrl, bytes.NewBuffer(fileContent))
+	if err != nil {
+		return err
+	}
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("upload failed with status code: %d", resp.StatusCode)
+	}
+
+	fmt.Printf("[LaunchDarkly] Uploaded %s to %s\n", filePath, name)
+	return nil
+}
+
+func initFlags(cmd *cobra.Command) {
+	cmd.Flags().String(apiKeyFlag, "", "The LaunchDarkly Observability API key")
```

Comment:
aside

@Vadman97 is this a regular LaunchDarkly API access token? earlier we talked about needing an API access token with a Writer role, or another role with write access to the LaunchDarkly project. I'm not sure from this description whether that's changed or not?

Reply:
the only key will be the regular access key that a customer uses with the CLI